src/spikeinterface/core/waveform_extractor.py

```diff
@@ -1863,53 +1893,56 @@ def save(self, **kwargs):
         self._save(**kwargs)
 
     def _save(self, **kwargs):
-        if self.format == "binary":
-            import pandas as pd
-
-            for ext_data_name, ext_data in self._extension_data.items():
-                if isinstance(ext_data, dict):
-                    with (self.extension_folder / f"{ext_data_name}.json").open("w") as f:
-                        json.dump(ext_data, f)
-                elif isinstance(ext_data, np.ndarray):
-                    np.save(self.extension_folder / f"{ext_data_name}.npy", ext_data)
-                elif isinstance(ext_data, pd.DataFrame):
-                    ext_data.to_csv(self.extension_folder / f"{ext_data_name}.csv", index=True)
-                else:
-                    try:
-                        with (self.extension_folder / f"{ext_data_name}.pkl").open("wb") as f:
-                            pickle.dump(ext_data, f)
-                    except:
-                        raise Exception(f"Could not save {ext_data_name} as extension data")
-        elif self.format == "zarr":
-            from .zarrrecordingextractor import get_default_zarr_compressor
-            import pandas as pd
-            import numcodecs
-
-            compressor = kwargs.get("compressor", None)
-            if compressor is None:
-                compressor = get_default_zarr_compressor()
-            for ext_data_name, ext_data in self._extension_data.items():
-                if ext_data_name in self.extension_group:
-                    del self.extension_group[ext_data_name]
-                if isinstance(ext_data, dict):
-                    self.extension_group.create_dataset(
-                        name=ext_data_name, data=[ext_data], object_codec=numcodecs.JSON()
-                    )
-                    self.extension_group[ext_data_name].attrs["dict"] = True
-                elif isinstance(ext_data, np.ndarray):
-                    self.extension_group.create_dataset(name=ext_data_name, data=ext_data, compressor=compressor)
-                elif isinstance(ext_data, pd.DataFrame):
-                    ext_data.to_xarray().to_zarr(
-                        store=self.extension_group.store, group=f"{self.extension_group.name}/{ext_data_name}", mode="a"
-                    )
-                    self.extension_group[ext_data_name].attrs["dataframe"] = True
-                else:
-                    try:
+        if not self.waveform_extractor.is_read_only():
```

Comment:
The diff of this is very difficult for me to catch up. Is it this a tab  of everything to the right?

If so, 

I think that you can get away by just doing an early return if the program is ready only.

That is:

```python
if self.waveform_extractor.is_read_only()
   return None
```

This will decrease nesting that at this level is very hard to read and keep track.

On the other hand, maybe this should ben assertion? That is, if a user wants to write to a folder that is read only maybe this is the point where we can tell them with a meanintgul assertion message?

What do you think?

Reply:
Yes it's a conditional load (only a tab separation).

While I agree that the diff iis harder to read, I think that logically it makes more sense to keep it this way. 
The logic is:
**If I can read it, then read it**

No?

Reply:
I think this makes sense if the scope of the scope of the function was simpler.

But here, there is like what, six levels of nesting below? 

I think that readability in the sense of removing one if block from your working memory thriumps faithful representation of logic.

A solution that followins from this line of reasoning is that maybe this if should go in in the function `save` and not here in `_save` as there the logic is expressed perfectly without hindering readability.

Reply:
done